src/test/ui/consts/const-eval/heap/alloc_intrinsic_nontransient.rs

```diff
@@ -0,0 +1,20 @@
+// run-pass
+#![feature(core_intrinsics)]
+#![feature(const_heap)]
+#![feature(const_raw_ptr_deref)]
+#![feature(const_mut_refs)]
+use std::intrinsics;
+
+const FOO: *const i32 = foo();
```

Comment:
It's super curious that this doesn't error with the same error that src/test/ui/consts/const-eval/heap/alloc_intrinsic_untyped.rs emits: "untyped pointer".

Reply:
Oh... I think I know why... `foo` is a function without any arguments, and is thus not actually called but treated as a constant itself and its body is thus evaluated directly: https://github.com/rust-lang/rust/blob/75f1db1102076e416e1154b241b4fc95c01c0d7b/compiler/rustc_mir/src/const_eval/machine.rs#L40-L44

Reply:
I guess we'll need to disable that optimization for function items now

Reply:
> I guess we'll need to disable that optimization for function items now

Why that?

Reply:
This test memoizes function alls to `foo()`, even though every call should be producing a new heap allocation. Otherwise all calls to [`Box::new_uninit()`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.new_uninit) will end up with the same heap allocation.

Reply:
Basically if you used

```rust
const fn foo() -> &'static i32 {
    const fn bar() -> *mut i32 {
        unsafe { intrinsics::const_allocate(4, 4) as *mut i32 }
    }
    unsafe {
        let i = bar();
        *i = 20;
        ...
    }
}
```

the `*i = 20` will fail to interpret because you'd be modifying interned memory

Reply:
Oh, lol... yeah, `const fn` without inputs are no longer pure with this change, That's kind of a big deal actually. We certainly need T-lang approval before stabilizing this.

Reply:
> Oh, lol... yeah, `const fn` without inputs are no longer pure with this change, That's kind of a big deal actually. We certainly need T-lang approval before stabilizing this.

But only if they return `&'static _` (as it's the memoizing of that result that's the problem here)?  Is that a useful escape hatch from purity semantics?  Or must `const fn foo() -> &'static str { "foo" }` always be pure?